presto-bigquery/src/main/java/com/facebook/presto/plugin/bigquery/BigQueryClient.java

```diff
@@ -0,0 +1,190 @@
+        return () -> Iterators.transform(datasets, this::addDataSetMappingIfNeeded);
+    }
+
+    protected Iterable<Table> listTables(DatasetId datasetId, TableDefinition.Type... types)
+    {
+        Set<TableDefinition.Type> allowedTypes = ImmutableSet.copyOf(types);
+        DatasetId bigQueryDatasetId = datasetIds.getOrDefault(datasetId, datasetId);
+        Iterable<Table> allTables = bigQuery.listTables(bigQueryDatasetId).iterateAll();
+        return StreamSupport.stream(allTables.spliterator(), false)
+                .filter(table -> allowedTypes.contains(table.getDefinition().getType()))
+                .collect(toImmutableList());
+    }
+
+    private Dataset addDataSetMappingIfNeeded(Dataset dataset)
+    {
+        DatasetId bigQueryDatasetId = dataset.getDatasetId();
+        DatasetId prestoDatasetId = DatasetId.of(bigQueryDatasetId.getProject(), bigQueryDatasetId.getDataset().toLowerCase(ENGLISH));
+        datasetIds.putIfAbsent(prestoDatasetId, bigQueryDatasetId);
+        return dataset;
+    }
+
+    protected TableId createDestinationTable(TableId tableId)
+    {
+        String project = viewMaterializationProject.orElse(tableId.getProject());
+        String dataset = viewMaterializationDataset.orElse(tableId.getDataset());
+        DatasetId datasetId = mapIfNeeded(project, dataset);
+        UUID uuid = randomUUID();
+        String name = format("_pbc_%s", randomUUID().toString().toLowerCase(ENGLISH).replace("-", ""));
+        return TableId.of(datasetId.getProject(), datasetId.getDataset(), name);
+    }
+
+    private DatasetId mapIfNeeded(String project, String dataset)
+    {
+        DatasetId datasetId = DatasetId.of(project, dataset);
+        return datasetIds.getOrDefault(datasetId, datasetId);
+    }
+
+    protected Table update(TableInfo table)
+    {
+        return bigQuery.update(table);
+    }
+
+    protected Job create(JobInfo jobInfo)
+    {
+        return bigQuery.create(jobInfo);
+    }
+
+    protected TableResult query(String sql)
+    {
+        try {
+            return bigQuery.query(QueryJobConfiguration.of(sql));
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new BigQueryException(BaseHttpServiceException.UNKNOWN_CODE, format("Failed to run the query [%s]", sql), e);
+        }
+    }
+
+    protected String createSql(TableId table, ImmutableList<String> requiredColumns, String[] filters)
+    {
+        String columns = requiredColumns.isEmpty() ? "*" :
+                requiredColumns.stream().map(column -> format("`%s`", column)).collect(joining(","));
+
+        String whereClause = createWhereClause(filters)
+                .map(clause -> "WHERE " + clause)
+                .orElse("");
+
+        return createSql(table, columns, filters);
+    }
+
+    // properly formatting SELECT statement
+    protected String createSql(TableId table, String formatedQuery, String[] filters)
```

Comment:
The name formatedQuery may not be precise, how about using columns or projection?

Reply:
revised to `requiredColumns` now